### DISCUSSION TITLE
The `on_init` step messes up colored syntax in Rust?

### DISCUSSION THREAD

**Discussion by ghost-in-the-zsh** [NvChad/NvChad#2970] [Q&A]
I've noticed that when working on Rust code, tokens like variable names, types, etc remain white:

![Screenshot from 2024-08-15 05-48-55](https://github.com/user-attachments/assets/f7c3cd64-ee25-4436-a726-00775a6c264d)

If I go to the `lspconfig.lua` module and comment out the `on_init` line, as shown below:

```lua
for _, lsp in ipairs(servers) do
  lspconfig[lsp].setup {
    -- on_init = on_init,
    on_attach = on_attach,
    capabilities = capabilities,
  }
end
```

then I see improved token/syntax coloring:

![Screenshot from 2024-08-15 05-49-50](https://github.com/user-attachments/assets/61fe68f1-80b1-4ed8-93ba-7ff53d49fc68)

Is this intended, a bug, or maybe a config issue? If intended, what should I do to keep the additional syntax coloring?

**PS:** I searched for color-related discussions, but the ones I found were about changing colors for look-and-feel, which is different from this one. They also predate the v2.5 release.

**Comment by siduck** [comment]
@ghost-in-the-zsh  TSInstall rust

**Reply by ghost-in-the-zsh** [reply]
@siduck It was already installed:
![Screenshot from 2024-08-15 06-02-37](https://github.com/user-attachments/assets/8ad6c8eb-3724-4adb-99da-72339594b023)

![Screenshot from 2024-08-15 06-03-25](https://github.com/user-attachments/assets/7a673791-ea99-46ab-a484-29303c691ca6)

**Reply by siduck** [reply]
hmm weird, but do knw that vars are colored in white in nvchad's base46, you can override this behavior.

Also the on_init func disables lsp semantic tokens highlighting

**Reply by ghost-in-the-zsh** [reply]
1. What's the recommended way to override the base46 behavior?
2. Why does `on_init` disable semantic token highlighting? What's the best way to override this behavior?

To clarify, by "best" and "recommended" I mean "best practices" and avoiding direct changes to the nvchad plugin to avoid causing conflicts, etc.

**Reply by siduck** [reply]
semantic tokens feel slow i.e it works after lsp is attached so i dont use it , as for override , use hl_override key

https://nvchad.com/docs/config/theming/

so place cursor on your variable and  run Inspect command  , then you will get to know the highlight group. Also ask stuff in our discord/tg/matrix, it'll be quicker than here!

**Reply by ghost-in-the-zsh** [reply]
> semantic tokens feel slow i.e it works after lsp is attached so i dont use it

I think I know what you mean. I notice it takes several seconds after opening `nvim` for the variables to get their syntax coloring to change. I've also noticed that some Python tokens don't get recognized as being language keywords and I'm not sure if it's related to this discussion or not. Does this sound familiar? For example, notice the `in` and `not` keywords within the `for` loop and `if` statement respectively:

![Screenshot from 2024-08-15 13-42-39](https://github.com/user-attachments/assets/e575205c-20bd-42da-b940-f263f75672cd)

FWIW, commenting `on_init` within the pylsp setup doesn't fix this.

> as for override , use hl_override key

I'll look into that. BTW, won't that cause conflicts when [this](https://github.com/NvChad/NvChad/issues/2938) gets released (if it hasn't been already :monocle_face:)? Or maybe I can/should already set it under the `M.base46` table?

> so place cursor on your variable and run Inspect command , then you will get to know the highlight group.

Interestingly, the variable shows up under the `rustFoldBraces` group. I thought it'd be grouped as some kind of identifier, like `InspectTree` shows it to be :thinking: :monocle_face:. In any case, I'll play around with it.

I wonder if there's anything that will automatically color different variables within the same scope with different colors. I remember having this kind of behavior in `Kate` when working on C++ and I liked it quite a bit. Are you aware of a way to get this kind of behavior or is this, AFAYK, not possible?

> Also ask stuff in our discord/tg/matrix, it'll be quicker than here!

I'll consider it. It was only a few days ago that I became aware of Matrix, so I'm kind of looking into that, too. Your TG seems to be a bridge from Matrix, which I read has some limitations re interactions, etc. In any case, I'll look closer into this.

**Reply by siduck** [reply]
i dont think variables can be colored differently , like if u got 5 vars in ur file, u want to color em all differently? u can ask in my discord server, many nvim  pros there

**Reply by ghost-in-the-zsh** [reply]
> i dont think variables can be colored differently , like if u got 5 vars in ur file, u want to color em all differently?

Yes, that. I'll see if I get around to asking some pro-nvimers. I haven't had much luck with what I've tried from the theming page, unfortunately. Still trying things out here and there...
